linmod/tests/test_eval.py

```diff
@@ -0,0 +1,345 @@
+                0.7,
+                0.4,
+                0.6,
+                0.5,
+                0.5,
+                0.2,
+                0.8,
+                0.3,
+                0.7,
+                0.4,
+                0.6,
+                0.5,
+                0.5,
+            ]
+        ),
+    )
+
+    correct_output = pl.DataFrame(
+        {
+            "fd_offset": [0, 0, 1, 1],
+            "division": [0, 1, 0, 1],
+            "mean_norm": [
+                np.abs(np.array([0.2, 0.3]) - 1 / (1 + 2)).mean()
+                + np.abs(np.array([0.8, 0.7]) - 2 / (1 + 2)).mean(),
+                np.abs(np.array([0.4, 0.5]) - 3 / (3 + 4)).mean()
+                + np.abs(np.array([0.6, 0.5]) - 4 / (3 + 4)).mean(),
+                np.abs(np.array([0.2, 0.3]) - 5 / (5 + 6)).mean()
+                + np.abs(np.array([0.8, 0.7]) - 6 / (5 + 6)).mean(),
+                np.abs(np.array([0.4, 0.5]) - 7 / (7 + 8)).mean()
+                + np.abs(np.array([0.6, 0.5]) - 8 / (7 + 8)).mean(),
+            ],
+        }
+    )
+
+    result = eval.proportions_mean_norm_per_division_day(
+        samples, data, p=1
+    ).collect()
+
+    assert_frame_equal(
+        result,
+        correct_output,
+        check_row_order=False,
+        check_column_order=True,
+        rtol=0,
+        atol=0,
+    )
+
+
+def test_proportions_L1_energy_score(
+    num_samples=100000, sample_variance=1.4, atol=0.05
+):
+    r"""
+    Test the estimate of the energy score of phi.
+
+    The setup is as follows:
+    - Generate fake data $Y_{tgl} \sim Uniform{0, ..., 99}$
+    - Denote true proportion as $\phi_{tgl} = \frac{Y_{tgl}}{Y_{tg \cdot}}$
+    - Generate fake "forecasts" $f_{tgl} \sim N(\phi_{tgl}, \sigma^2)$
+    - Check $\sum_{t, g} E[ || f_{tg} - \phi_{tg} ||_2 ] - \frac{1}{2} E[ || f_{tg} - \f_{tg}' ||_2 ]$
+      as reported by `eval.proportions_energy_score`
+
+    Note that our "forecasts" are not actually proportions, but independent normal
+    random variables. This is so that the quantity can be computed analytically.
+    """
+
+    rng = default_rng()
+
+    NUM_DAYS = 4
+    NUM_DIVISIONS = 3
+    NUM_LINEAGES = 2
+
+    samples, data = _generate_fake_samples_and_data(
+        rng,
+        num_days=NUM_DAYS,
+        num_divisions=NUM_DIVISIONS,
+        num_lineages=NUM_LINEAGES,
+        num_samples=num_samples,
+        sample_variance=sample_variance,
+    )
+
+    # Because we use L1 norm, term 1 is equal to the sum of each component's MAE from
+    # its mean.
+    # The MAE of a normal random variable from its mean is $\sigma \sqrt{\frac{2}{\pi}}$.
+
+    term1 = (
+        np.sqrt(2 * sample_variance / np.pi)
+        * NUM_DAYS
+        * NUM_DIVISIONS
+        * NUM_LINEAGES
+    )
+
+    # Because we use L1 norm, term 2 is equal to the sum of each component's MAE from
+    # an independent copy of itself.
+    # The MAE of a normal random variable from an independent copy is $\frac{2\sigma}{\sqrt{\pi}}$.
```

Comment:
One of these days I'm going to learn what the energy score is actually doing, but this seems like a neat analytical way to scientifically test correctness.